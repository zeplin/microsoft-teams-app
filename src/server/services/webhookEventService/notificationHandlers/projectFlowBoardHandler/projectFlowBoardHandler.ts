import { ProjectFlowBoardBuiltEvent, WebhookEvent } from "@zeplin/sdk";

import { GroupingKeyParams, NotificationHandler } from "../NotificationHandler";
import { MessageCard, commonTeamsCard } from "../teamsCardTemplates";
import { md } from "../md";
import { getRedirectURLForZeplinApp, getWebAppURL } from "../zeplinURL";

type Event = ProjectFlowBoardBuiltEvent;

class ProjectFlowBoardHandler extends NotificationHandler<Event> {
    delay = 0;

    private getText(event: Event): string {
        const {
            context: {
                project: {
                    name: projectName
                }
            },
            actor: {
                user: {
                    username
                }
            }
        } = event;
        return md`${username} built the first flow for ${projectName}. ⤴️`;
    }

    private getWebappURL(event: Event): string {
        const {
            context: {
                project: {
                    id: projectId
                }
            }
        } = event;

        return getWebAppURL(`project/${projectId}`);
    }

    private getZeplinAppURI(event: Event): string {
        const {
            context: {
                project: {
                    id: projectId
                }
            }
        } = event;

        const resource = "project";
        const searchParams = {
            pid: projectId
        };

        return getRedirectURLForZeplinApp(resource, searchParams);
    }

    getTeamsMessage(events: Event[]): MessageCard {
        const [event] = events;
        return commonTeamsCard({
            text: this.getText(event),
            links: [{
                title: "Open in App",
                url: this.getZeplinAppURI(event)
            }, {
                title: "Open in Web",
                url: this.getWebappURL(event)
            }],
            section: {
                text: "Flows in Zeplin provide a bird's-eye view of how screens in a project connect. [Learn more.](https://zpl.io/flows)"
            }
        });
    }

    shouldHandleEvent(event: WebhookEvent): event is ProjectFlowBoardBuiltEvent {
        return event.action === "built";
    }

    // A unique grouping key so that it won't be grouped with any other events
    getGroupingKey({ deliveryId, webhookId }: GroupingKeyParams<ProjectFlowBoardBuiltEvent>): string {
        return `${webhookId}:${deliveryId}`;
    }
}

export const projectFlowBoardHandler = new ProjectFlowBoardHandler();
