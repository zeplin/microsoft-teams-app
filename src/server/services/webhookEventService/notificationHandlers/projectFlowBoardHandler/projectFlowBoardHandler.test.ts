import { ProjectFlowBoardBuiltEvent } from "@zeplin/sdk";

import { projectFlowBoardHandler } from "./projectFlowBoardHandler";

type GetDummyEventParams = {
    action?: string;
    flowBoardId?: string;
}

function getDummyEvent({
    action = "built",
    flowBoardId = "flowBoardId"
}: GetDummyEventParams = {}): ProjectFlowBoardBuiltEvent {
    return {
        action,
        context: {
            project: {
                id: "projectId",
                name: "projectName"
            }
        },
        actor: {
            user: {
                username: "testUsername"
            }
        },
        resource: {
            id: flowBoardId,
            type: "FlowBoard",
            data: {
                id: flowBoardId
            }
        }
    } as ProjectFlowBoardBuiltEvent;
}

describe("projectBoardHandler", () => {
    describe("getTeamsMessage method", () => {
        beforeAll(() => {
            jest.spyOn(Math, "random").mockReturnValue(0);
        });

        describe("for created notification", () => {
            it("should match snapshot when there is only 1 event", () => {
                expect(
                    projectFlowBoardHandler.getTeamsMessage([getDummyEvent()])
                ).toMatchSnapshot();
            });
        });
    });
});
