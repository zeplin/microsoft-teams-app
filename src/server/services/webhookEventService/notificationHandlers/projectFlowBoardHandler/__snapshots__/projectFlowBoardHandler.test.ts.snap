// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`projectBoardHandler getTeamsMessage method for created notification should match snapshot when there is only 1 event 1`] = `
{
  "@type": "MessageCard",
  "@context": "https://schema.org/extensions",
  "text": "testUsername built the first flow for projectName. ⤴️",
  "sections": [
    {
      "text": "Flows in Zeplin provide a bird's-eye view of how screens in a project connect. [Learn more.](https://zpl.io/flows)",
      "images": []
    }
  ],
  "potentialAction": [
    {
      "@type": "OpenUri",
      "name": "Open in App",
      "targets": [
        {
          "os": "default",
          "uri": "https://msteams-app.zeplin.io/zeplin/app-redirect?uri=zpl%3A%2F%2Fproject%3Fpid%3DprojectId%26ref%3Dmicrosoft-teams"
        }
      ]
    },
    {
      "@type": "OpenUri",
      "name": "Open in Web",
      "targets": [
        {
          "os": "default",
          "uri": "https://app.zeplin.io/project/projectId?ref=microsoft-teams"
        }
      ]
    }
  ]
}
`;
